src/viam/sdk/components/encoder/encoder.hpp

```diff
@@ -0,0 +1,114 @@
+/// @file components/encoder/encoder.hpp
+///
+/// @brief Defines a `Encoder` component.
+#pragma once
+
+#include <string>
+
+#include <viam/api/component/encoder/v1/encoder.pb.h>
+
+#include <viam/sdk/common/proto_type.hpp>
+#include <viam/sdk/common/utils.hpp>
+#include <viam/sdk/config/resource.hpp>
+#include <viam/sdk/registry/registry.hpp>
+#include <viam/sdk/resource/resource_manager.hpp>
+
+namespace viam {
+namespace sdk {
+
+/// @defgroup Encoder Classes related to the `Encoder` component.
+
+/// @class EncoderSubtype
+/// @brief Defines a `ResourceSubtype` for the `Encoder` component.
+/// @ingroup Encoder
+class EncoderSubtype : public ResourceSubtype {
+   public:
+    std::shared_ptr<ResourceServerBase> create_resource_server(
+        std::shared_ptr<ResourceManager> manager) override;
+    std::shared_ptr<ResourceBase> create_rpc_client(std::string name,
+                                                    std::shared_ptr<grpc::Channel> chan) override;
+    EncoderSubtype(const google::protobuf::ServiceDescriptor* service_descriptor)
+        : ResourceSubtype(service_descriptor){};
+};
+
+/// @class Encoder encoder.hpp "components/encoder/encoder.hpp"
+/// @brief An encoder is a device that is hooked up to motors to report a position
+/// @ingroup Encoder
+///
+/// This acts as an abstract base class to be inherited from by any drivers representing
+/// specific encoder implementations. This class cannot be used on its own.
+class Encoder : public ComponentBase {
+   public:
+    /// @enum position_type
+    enum position_type {
+        // Unspecified position type
+        UNSPECIFIED,
+        // Provided by incremental encoders
+        TICKS_COUNT,
+        // Provided by absolute encoders
+        ANGLE_DEGREES
+    };
+
+    /// @struct position
+    /// @brief reported position.
+    struct position {
+        float value;
+        position_type type;
+    };
+
+    /// @struct properties
+    /// @brief Encodes the supported modes of this encoder
+    struct properties {
```

Comment:
Are all states of this struct meaningful / possible? Or only two?

Reply:
I don't fully understand this question. My (elementary) understanding of encoders is that all four permutations of these booleans are possible values.

Reply:
Yeah, I just didn't know, and I try to avoid types that have invalid states. If only one of these bools could meaningfully be true at once I would have suggested another approach.

It feels a little odd to me still that you have the enum, but then you also have a bool associated with each value of the enum. Just throwing out some ideas but maybe you could return `std::vector<position_type>` where only the supported position types were returned? Or `std::map<position_type, bool>`? As it stands you will need to change the size of this structure if ever a new `position_type` were added. And you need to know the name of two symbols to see if a certain `position_type` is supported, rather than just knowing the enumeration names.

Reply:
I agree that this is confusing and a bit unfortunate. 

I am hesitant to modify this too much because I suspect it was done this way to allow Viam to easily modify the `properties` struct going forward. Dropping to `vec<position_type>` would block us in and would make us different from the other sdks.